### PR TITLE
Feature/decrypt

### PR DESCRIPTION
**Pull request by norling**
Finally, the decryption module! The final PR from me

### CODE REVIEW

File: decrypt/decrypt.go [hunk 1]
```diff
@@ -1,42 +1,213 @@
...
 	log "github.com/sirupsen/logrus"
 )
 
 // Help text and command line flags.
 
+// Usage text that will be displayed as command line help text when using the
+// `help decrypt` command
 var Usage = `
-USAGE: %s decrypt -key <private-key-file> [file(s)]
+USAGE: %s decrypt (-createKey <name>) -key <private-key-file> [file(s)]
```
Comment (jbygdell) [inline decrypt/decrypt.go:23]:
```suggestion
USAGE: %s decrypt -key <private-key-file> [file(s)]
```
CreateKey has no use-case in a decryption scenario

File: decrypt/decrypt.go [hunk 2]
```diff
@@ -1,42 +1,213 @@
...
 
 Decrypt: Encrypts files from the Sensitive Data Archive (SDA) with the provided
-         private key.
+         private key. If the private key is encrypted, the password can be
+		 supplied in the DECRYPT_PASSWORD environment variable, or at the
+		 interactive password prompt.
```
Comment (jbygdell) [inline decrypt/decrypt.go:28]:
```suggestion
         private key. If the private key is encrypted, the password can be
	 supplied in the DECRYPT_PASSWORD environment variable, or at the
	 interactive password prompt.
```

File: decrypt/decrypt.go [hunk 3]
```diff
@@ -1,42 +1,213 @@
...
-var privateKey = Args.String("key", "",
+var privateKeyFile = Args.String("key", "",
 	"Private key to use for decrypting files.")
 
-// Main decryption function
-func Decrypt(args []string) {
+// Decrypt takes a set of arguments, parses them, and attempts to eiher create a
+// crypt4gh key pair (if -keyName is set) or decrypt the given data files with
+// the given private key file (otherwise).
+func Decrypt(args []string) error {
```
Comment (jbygdell) [inline decrypt/decrypt.go:47]:
```suggestion
// Decrypt takes a set of arguments, parses them, and attempts to decrypt the
// given data files with the given private key file..
func Decrypt(args []string) error {
```

File: decrypt/decrypt.go [hunk 4]
```diff
@@ -1,42 +1,213 @@
...
+	}
+
+	var privateKey *[32]byte
+
+	// try reading private key without password
+	privateKey, err = readPrivateKey(*privateKeyFile, "")
+	if err != nil {
+
+		// if there was an error, try again with the password
+		password, err := getPassword("DECRYPT_PASSWORD")
```
Comment (jbygdell) [inline decrypt/decrypt.go:77]: Better ENV name for the private key password might be in order

File: decrypt/decrypt_test.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+	// Check content of the decrypted file
+	inFile, err := os.Open(decryptedFile)
+	if err != nil {
+		log.Fatalf("Couldn't open decrypted file %s for content checking", decryptedFile)
+	}
+	fileData, err := ioutil.ReadAll(inFile)
+	if err != nil {
+		log.Fatal("Couldn't read decrypted filedata for content checking")
+	}
```
Comment (jbygdell) [inline decrypt/decrypt_test.go:163]: Use asserts in tests instead of log.fatal

File: decrypt/decrypt_test.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	encryptArgs := []string{"sda-cli", "-key", fmt.Sprintf("%s.pub.pem", testKeyFile), suite.testFile.Name()}
+	err = encrypt.Encrypt(encryptArgs)
+	if err != nil {
+		log.Fatalf("couldn't encrypt file for decryption test: %s", err)
+	}
+	err = os.Chdir(cwd)
+	if err != nil {
+		log.Fatal("could not return from test directory")
+	}
```
Comment (jbygdell) [inline decrypt/decrypt_test.go:136]: Use asserts in tests instead of log.fatal

File: decrypt/decrypt_test.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+	// generate key files
+	err := createKey.GenerateKeyPair(testKeyFile, "")
+	if err != nil {
+		log.Fatalf("couldn't generate testing key pair: %s", err)
+	}
+	// and read the private key
+	privateKey, err := readPrivateKey(fmt.Sprintf("%s.sec.pem", testKeyFile), "")
+	if err != nil {
+		log.Fatalf("couldn't read test key: %s", err)
+	}
```
Comment (jbygdell) [inline decrypt/decrypt_test.go:116]: Use asserts in tests instead of log.fatal

File: decrypt/decrypt_test.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+func (suite *DecryptTests) TestreadPrivateKey() {
+
+	testKeyFile := filepath.Join(suite.tempDir, "testkey")
+
+	// generate key files
+
+	err := createKey.GenerateKeyPair(testKeyFile, "")
+	if err != nil {
+		log.Fatalf("couldn't generate testing key pair: %s", err)
+	}
```
Comment (jbygdell) [inline decrypt/decrypt_test.go:68]: Use asserts in tests instead of log.fatal

File: decrypt/decrypt_test.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+	// ... create some content ...
+	suite.fileContent = []byte("This is some fine content right here.")
+
+	// ... and write the known content to it
+	err = ioutil.WriteFile(suite.testFile.Name(), suite.fileContent, 0600)
+	if err != nil {
+		log.Fatalf("failed to write to testfile: %s", err)
+	}
```
Comment (jbygdell) [inline decrypt/decrypt_test.go:52]: Use asserts in tests instead of log.fatal

**Review by jbygdell — CHANGES_REQUESTED** [review]

File: decrypt/decrypt.go (hunk 4 above)
Comment (norling) [reply decrypt/decrypt.go:76]: YES! I couldn't come up with one, so I just used something descriptive that wasn't tied to SDA or BP... any ideas? or should I make a flag for the ENV?

File: decrypt/decrypt_test.go (hunk 9 above)
Comment (norling) [reply decrypt/decrypt_test.go:52]: I used Log.Fatal when there were failures unrelated to the tests, but I can replace them with asserts instead :)

File: decrypt/decrypt.go (hunk 4 above)
Comment (jbygdell) [reply decrypt/decrypt.go:76]: C4GH_PASSWORD perhaps

Comment (norling) [reply decrypt/decrypt.go:76]: sure! :)

**Review by jbygdell — APPROVED** [review]

**Review by aaperis — COMMENTED** [review]
looks good! maybe add 
`"createKey": {createKey.Args, createKey.Usage, createKey.ArgHelp},`
to `main.go` ?

**Comment by norling** [comment]
Great catch! Added to main, and fixed help text.

File: create_key/create_key.go [hunk 10]
```diff
@@ -29,7 +29,9 @@ createKey: Creates a crypt4gh encryption key pair, and saves it to
 
 // ArgHelp is the suffix text that will be displayed after the argument list in
 // the module help.
-var ArgHelp = ``
+var ArgHelp = `
+  [basename]
+        The basename of the keyfiles to generate`
```
Comment (aaperis) [inline create_key/create_key.go:34]: is [basename] meant for `<name>` ?

```
$ ./sda-cli help createKey

USAGE: ./sda-cli createKey <name> (-outdir <dirname>)
```

**Pull request merged**